hack/builder/Dockerfile

```diff
@@ -45,6 +46,16 @@ RUN dnf install -y dnf-plugins-core && \
         skopeo && \
     dnf clean -y all
 
+RUN if test "${ARCH}" = amd64; then \
```

Comment:
@brianmcarey have we agreed on the crosscomplation of s390x?

Reply:
How do you mean agreed on? Do you mean regarding the approach? 

I believe its following the same approach as what is used for the arm64 cross compilation. 
We have a presubmit prow job that checks the s390x cross compilation.

Reply:
Are we cross-compiling arm? I would then expect `gcc-aarch64-linux-gnu` in the Dockerfile

Reply:
Arm has a pre-built cross-compile chain that is downloaded and used instead. So the way the cross-compile toolchain is provided is different between arm and s390x.
See: https://github.com/kubevirt/kubevirt/blob/main/third_party/toolchains/toolchains.bzl

Reply:
Yes we are - https://github.com/kubevirt/project-infra/blob/a42b4b4c30fd796fe2b2f4c3d0c89896a6f9a83e/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml#L489C1-L508C27

I would expect `gcc-aarch64-linux-gnu` too but I don't see it included in the builder publish.

@zhlhahaha do you know how this works for arm64?

Reply:
As I mentioned in the community meeting, we could create a new image that depends on the buider but it contains the additional cross-compilation tools. The advantages of this approach are:
  * for builds on x86 it doesn't need to pull the extra dependencies
  * we can have a better split and ownership for the maintenance of the new image

Reply:
Yes sounds good to me. There are a couple of drawbacks but they are minor. 
- In the rare occasion, where a contributor wants to cross compile for s390x locally - they will have to know to use the other image. This probably should be documented. 
- Extra CI jobs for publishing and testing this new builder image - so some extra overhead here but it shouldn't change much.

Reply:
> Yes sounds good to me. There are a couple of drawbacks but they are minor.
> 
> * In the rare occasion, where a contributor wants to cross compile for s390x locally - they will have to know to use the other image. This probably should be documented.

I think this could be detected in the build script and we can print a warning message

> * Extra CI jobs for publishing and testing this new builder image - so some extra overhead here but it shouldn't change much.

Yes, this is the main question and my main worry. Do we want this?

Reply:
Previous discussion about this: https://github.com/kubevirt/kubevirt/pull/10490#issuecomment-1813008072 and comments around it.

It's not up to me to make a call, but there's certainly a trade-off to be made between increasing the size of the builder image by a fairly noticeable amount for everyone, including those who will never care about cross-compilation, and increasing the complexity of the build system and CI jobs.

In my opinion it's kinda silly to worry about adding <1GB for an image that is updated very infrequently, when new k8s images that are 10GB+ in size need to be pulled by everyone all the time. But again, it's not my call :)

Reply:
I did some research here and the arm64 toolchain has not been updated in 2 years. This means x86 is currently using gcc 11.4.1, whereas aarch64 is using gcc 11.3. So it might be beneficial to switch arm to the same approach, to ensure matching compiler versions across all arches.

/cc @zhlhahaha 

Additionally here are my thoughts to implementing this:
- Currently all build scripts invoked by make run inside dockerized enviroment from the start -> No ability to switch image in build scripts
- Add logic in `hack/dockerized` to check if running on x86_64 and `BUILD_ARCH` contains at least 1 arch other than x86_64
  - When cross compile is needed, default to cross-compile image (User can still override image)
  - When no cross-compile, default to normal builder image
  - Highest priority is always user choosen container image

WDYT @alicefr @brianmcarey @andreabolognani

Reply:
> This means x86 is currently using gcc 11.4.1, whereas aarch64 is using gcc 11.3.

As I did not follow the discussion, so I only answer this question. Currently, we only provide following version gun-toolchain, and gcc11.4.1 is not in the list:
[Downloads: 13.2.Rel1 October 30, 2023](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel1a)
[Downloads: 12.3.Rel1 July 28, 2023](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel2a)
[Downloads: 12.2.MPACBTI-Rel1 March 22, 2023](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel3a)
[Downloads: 12.2.Rel1 December 22, 2022](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel4a)
[Downloads: 12.2.MPACBTI-Bet1 September 06, 2022](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel5a)
[Downloads: 11.3.Rel1 August 08, 2022](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel6a)
[Downloads: 11.2-2022.02 February 15, 2022](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads#panel7a)

For more detail please refer for https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads

Reply:
Definitely, we should use the same method for ARM and s390x. I would be curious to see if we move out the ARM crossbuild toolchain of how much the builder image size decreases. 
It is true that k8s images are very large (we should also do something about that). However, we don't pull them for the unit tests and building.
After saying that, this is a proposal. If the CI work is too complex, then we can keep it in the same Dockerfile.

Reply:
I was definitely planning to switch the aarch64 cross-compiler to use the same method. I should still have the patches around, I was just waiting for this work to land so that they would be trivial to justify.

Note that currently the aarch64 toolchain is not part of the builder image, but rather is downloaded and unpacked on demand when needed. That's a nice characteristic, and I tried to achieve the same for the s390x toolchain, but after spending several days in a row fighting with bazel only to get nowhere close to the goal I simply gave up and accepted the compromise implemented here.

I can imagine someone better versed in bazel being able to eventually manage what I couldn't, but I don't think either me or Jan are the right people for the task, and more importantly I don't think it should block this work. The current trade-off is not ideal, but still falls completely within the bounds of what's reasonable IMO.

Reply:
I think i will try out my idea on how to achieve it in a separate image with your changes @andreabolognani and then raise a draft pr were we can discuss/review it more easily. That way we have some code to look at along to the theory and i can also put the cross-compile functionality from this PR to over there, so this big PR gets just a little bit more slim.

Reply:
@jschintag do you know how much the cross-compilation dependencies add to the original builder image?

Reply:
@alicefr see https://github.com/kubevirt/kubevirt/pull/10490#issuecomment-1813008072 mentioned earlier. It's ~400MB extra per architecture. The data is from a few months ago, but I don't expect the numbers to have changed significantly in the meantime.

Reply:
For a shared cross-compile concept, i have created and tested it here: https://github.com/kubevirt/kubevirt/compare/main...jschintag:kubevirt:cross-compile

This would build a separate cross-compile image based on the builder image.
Then depending on if there is a need to target a cross compile build, `hack/dockerized` will use the cross-compile image as long as no specific build image is specified.

WDYT @alicefr @brianmcarey ?

Reply:
@brianmcarey what are your thoughts? I think both are doable. If we can avoid to pull 200-400MB it won't be a bad idea, but still if this doesn't over complicate the entire setup

Reply:
Yes I think avoiding the extra image size will be nice as the majority of the contributors will not be trying to cross-compile so that will avoid a lot of waste and it still allows us to test it in CI so it sounds good to me!

Reply:
I have created a separate PR https://github.com/kubevirt/kubevirt/pull/11991 for the cross-compile, let me tag all the participants over there so we can continue/resolve the discussion over there.